docs/man/man3/AEFileWriter.3
.TH "AEFileWriter" 3 "Mon Mar 18 2024 18:44:24" "Version v0.0.8.5a" "ArtyK's Console Engine" \" -*- nroff -*-
.ad l
.nh
.SH NAME
AEFileWriter \- ArtyK's Engine File Writer; umm, it writes data to given file\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fR#include <AEFileWriter\&.hpp>\fP
.PP
Inherits \fB__AEModuleBase< AEFileWriter >\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBAEFileWriter\fP (const std::string_view filename, const \fBcint\fP flags=0, const \fBullint\fP af_interval=\fBAEFW_DEFAULT_AUTOFLUSH_INTERVAL\fP)"
.br
.RI "Class constructor -- creates the \fBAEFileWriter\fP instance and opens the file\&. "
.ti -1c
.RI "\fBAEFileWriter\fP (void) noexcept"
.br
.RI "Class constructor -- creates the instance but doesn't do anything\&. "
.ti -1c
.RI "\fBAEFileWriter\fP (\fBAEFileWriter\fP &&fw) noexcept"
.br
.RI "Move constructor\&. "
.ti -1c
.RI "\fBAEFileWriter\fP (const \fBAEFileWriter\fP &)=delete"
.br
.RI "Deleted copy constructor\&. "
.ti -1c
.RI "\fBAEFileWriter\fP & \fBoperator=\fP (const \fBAEFileWriter\fP &)=delete"
.br
.RI "Deleted copy assignment operator\&. "
.ti -1c
.RI "\fB~AEFileWriter\fP (void) noexcept"
.br
.RI "Default Destructor\&. "
.ti -1c
.RI "\fBcint\fP \fBopenFile\fP (const std::string_view str, const \fBcint\fP flags=0, const \fBullint\fP af_interval=\fBAEFW_DEFAULT_AUTOFLUSH_INTERVAL\fP)"
.br
.RI "Open(and create) file with given name and flag\&. "
.ti -1c
.RI "\fBcint\fP \fBflushFile\fP (void) const noexcept"
.br
.RI "Flushes the opened file\&. "
.ti -1c
.RI "\fBcint\fP \fBcloseFile\fP (void) noexcept"
.br
.RI "Closes the currently opened file\&. "
.ti -1c
.RI "template<typename T > \fBcint\fP \fBwrite\fP (const T &var, const size_t datasz=0, const bool useAutoFlush=\fBAEFW_DEFAULT_AUTOFLUSH\fP) noexcept"
.br
.RI "Generic function for writing data, invoking proper write functions for built-in types\&. "
.ti -1c
.RI "\fBcint\fP \fBwriteString\fP (const std::string &str, const bool includeNull=false, const bool useAutoFlush=\fBAEFW_DEFAULT_AUTOFLUSH\fP) noexcept"
.br
.RI "Write std::string, as is, to file, and flush if necessary\&. "
.ti -1c
.RI "\fBcint\fP \fBwriteString\fP (const std::vector< char > &str, const bool useAutoFlush=\fBAEFW_DEFAULT_AUTOFLUSH\fP) noexcept"
.br
.RI "Write std::vector<char> (string), as is, to file, and flush if necessary\&. "
.ti -1c
.RI "\fBcint\fP \fBwriteString\fP (const char *const cdata, const bool includeNull=false, const bool useAutoFlush=\fBAEFW_DEFAULT_AUTOFLUSH\fP) noexcept"
.br
.RI "Write c-string, as is, to file, and flush if necessary\&. "
.ti -1c
.RI "\fBcint\fP \fBwriteInt\fP (const \fBllint\fP num, const bool useAutoFlush=\fBAEFW_DEFAULT_AUTOFLUSH\fP) noexcept"
.br
.RI "Writes the (signed) integer as text to open file\&. "
.ti -1c
.RI "\fBcint\fP \fBwriteUInt\fP (const \fBullint\fP num, const bool useAutoFlush=\fBAEFW_DEFAULT_AUTOFLUSH\fP) noexcept"
.br
.RI "Writes the (unsigned) integer as text to open file\&. "
.ti -1c
.RI "template<typename T > 
.br
requires (std::is_floating_point<T>::value == true)\fBcint\fP \fBwriteFloat\fP (const T num, const bool useAutoFlush=\fBAEFW_DEFAULT_AUTOFLUSH\fP) noexcept"
.br
.RI "Writes the float value as text to the opened file\&. "
.ti -1c
.RI "\fBcint\fP \fBwriteBool\fP (const bool num, const bool useAutoFlush=\fBAEFW_DEFAULT_AUTOFLUSH\fP) noexcept"
.br
.RI "Writes the boolean to file as text (true/false)\&. "
.ti -1c
.RI "\fBcint\fP \fBwriteChar\fP (const char c, const bool useAutoFlush=\fBAEFW_DEFAULT_AUTOFLUSH\fP) noexcept"
.br
.RI "Writes the single (ascii) character to the file\&. "
.ti -1c
.RI "\fBcint\fP \fBwriteByte\fP (const unsigned char cdata, const bool useAutoFlush=\fBAEFW_DEFAULT_AUTOFLUSH\fP) noexcept"
.br
.RI "Writes one byte of data to the file\&. "
.ti -1c
.RI "\fBcint\fP \fBwriteBytes\fP (const void *const cdata, const std::size_t dsize, const bool useAutoFlush=\fBAEFW_DEFAULT_AUTOFLUSH\fP) noexcept"
.br
.RI "Write a stream of bytes to file, from pointer\&. "
.ti -1c
.RI "\fBcint\fP \fBwriteBytes\fP (const std::vector< unsigned char > &cdata, const bool useAutoFlush=\fBAEFW_DEFAULT_AUTOFLUSH\fP) noexcept"
.br
.RI "Write a stream of bytes to file, from std::vector<unsigned char>\&. "
.ti -1c
.RI "\fBcint\fP \fBwriteData_ptr\fP (const void *const cdata, const std::size_t dcount, const std::size_t dsize=sizeof(char), const bool useAutoFlush=\fBAEFW_DEFAULT_AUTOFLUSH\fP) noexcept"
.br
.RI "Write binary data, as is, to file, and flush if necessary\&. "
.ti -1c
.RI "template<typename T > \fBcint\fP \fBwriteData_ref\fP (const T &cdata, const std::size_t tsize=sizeof(T), const bool useAutoFlush=\fBAEFW_DEFAULT_AUTOFLUSH\fP) noexcept"
.br
.RI "Write binary data, as is, to file, and flush if necessary\&. "
.ti -1c
.RI "bool \fBisOpen\fP (void) const noexcept"
.br
.RI "Checks if a file is open by this file-writer\&. "
.ti -1c
.RI "bool \fBisClosed\fP (void) const noexcept"
.br
.RI "Checks if a file isn't open by this file-writer\&. "
.ti -1c
.RI "\fBllint\fP \fBgetFileSize\fP (void) const noexcept"
.br
.RI "Returns size of the file in bytes\&. "
.ti -1c
.RI "std::string \fBgetFullFileName\fP (void) const noexcept"
.br
.RI "Returns the full name of currently open file, including the relative path, if such was provided\&. "
.ti -1c
.RI "std::string \fBgetRelativePath\fP (void) const"
.br
.RI "Returns the relative path of the opened file\&. "
.ti -1c
.RI "std::string \fBgetFileName\fP (void) const"
.br
.RI "Returns the name of the opened file, devoid of the path\&. "
.ti -1c
.RI "std::string \fBgetFullPath\fP (void) const"
.br
.RI "Returns the full, absolute path of the opened file\&. "
.ti -1c
.RI "std::FILE * \fBgetFilePtr\fP (void) const noexcept"
.br
.RI "Returns the file pointer of this file-writer\&. "
.ti -1c
.RI "\fBllint\fP \fBgetCursorPos\fP (void) const noexcept"
.br
.RI "Returns current write cursor position\&. "
.ti -1c
.RI "int \fBsetCursorPos\fP (const \fBllint\fP pos, const int origin=SEEK_CUR) const noexcept"
.br
.RI "Sets read cursor position to pos from origin\&. "
.ti -1c
.RI "\fBullint\fP \fBgetTotalWrites\fP (void) const noexcept"
.br
.RI "Returns total write requests made to file\&. "
.ti -1c
.RI "std::size_t \fBgetLastWrittenAmount\fP (void) const noexcept"
.br
.RI "Gets the last written amount of bytes from the opened file\&. "
.ti -1c
.RI "constexpr const \fBullint\fP \fBgetModuleID\fP (void) noexcept"
.br
.RI "Get the \fBid number of the instance\fP of the module\&. "
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static \fBullint\fP \fBgetModuleAmount\fP (void) noexcept"
.br
.RI "Get the \fBamount of the currently-alive instances of module T\fP\&. "
.ti -1c
.RI "static constexpr const std::string_view \fBgetModuleName\fP (void) noexcept"
.br
.RI "Get the \fBname of the module\fP\&. "
.ti -1c
.RI "static constexpr const \fBullint\fP \fBgetInstanceCounter\fP (void) noexcept"
.br
.RI "Get the current value of the \fBinstance counter\fP of the module\&. "
.in -1c
.SS "Public Attributes"

.in +1c
.ti -1c
.RI "\fBullint\fP \fBm_ullFlushInterval\fP"
.br
.RI "Interval in write operations before automatic flush operation\&. "
.in -1c
.SS "Protected Attributes"

.in +1c
.ti -1c
.RI "const \fBullint\fP \fBm_ullModuleId\fP"
.br
.RI "The ID number of the current module\&. "
.in -1c
.SS "Static Protected Attributes"

.in +1c
.ti -1c
.RI "static std::atomic< std::size_t > \fBm_ullModulesAlive\fP"
.br
.RI "Amount of the currently-alive (not destroyed) instances of the module\&. "
.ti -1c
.RI "static std::atomic< \fBullint\fP > \fBm_ullInstanceCounter\fP"
.br
.RI "Amount of the instances of the module ever created during the program's runtime\&. "
.in -1c
.SS "Private Member Functions"

.in +1c
.ti -1c
.RI "void \fBautoFlush\fP (void) noexcept"
.br
.RI "Checks for and performs the auto-flush operation if required\&. "
.ti -1c
.RI "\fBREGISTER_MODULE\fP (\fBAEFileWriter\fP)"
.br
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "std::string \fBm_sFilename\fP"
.br
.RI "Full filename and relative path\&. "
.ti -1c
.RI "std::atomic< \fBullint\fP > \fBm_ullTotalWrites\fP"
.br
.RI "Counter for total write operations for file\&. "
.ti -1c
.RI "std::size_t \fBm_szLastWrittenAmount\fP"
.br
.RI "The amount of written bytes during last operation\&. "
.ti -1c
.RI "std::FILE * \fBm_fpFilestr\fP"
.br
.RI "Object for file writing\&. "
.ti -1c
.RI "\fBcint\fP \fBm_cFlags\fP"
.br
.RI "Flags that were used to open the file\&. "
.in -1c
.SH "Detailed Description"
.PP 
ArtyK's Engine File Writer; umm, it writes data to given file\&. 

It is a wrapper around the C's FILE api, for speed and convenience\&. It can write strings, bools, ints, and floats, both as raw bytes and formatted to text
.PP
Just create it and dump the gigabytes of data to your files\&. Hungarian notation is fw\&. (m_fwMyFileWriter) Flags start with AEFW_ 
.PP
\fBWarning\fP
.RS 4
This is not thread safe! 
.RE
.PP

.PP
Definition at line \fB108\fP of file \fBAEFileWriter\&.hpp\fP\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "AEFileWriter::AEFileWriter (const std::string_view filename, const \fBcint\fP flags = \fR0\fP, const \fBullint\fP af_interval = \fR\fBAEFW_DEFAULT_AUTOFLUSH_INTERVAL\fP\fP)\fR [explicit]\fP"

.PP
Class constructor -- creates the \fBAEFileWriter\fP instance and opens the file\&. 
.PP
\fBSee also\fP
.RS 4
\fBAEFileWriter::openFile()\fP
.RE
.PP
\fBParameters\fP
.RS 4
\fIfilename\fP Name of the file, with extension
.br
\fIflags\fP Flags for file opening; look up AEFW_FLAG_* for more info
.br
\fIaf_interval\fP interval in file writes between automatic file flushing 
.RE
.PP

.PP
Definition at line \fB12\fP of file \fBAEFileWriter\&.cpp\fP\&.
.PP
References \fBopenFile()\fP\&.
.SS "AEFileWriter::AEFileWriter (void)\fR [inline]\fP, \fR [noexcept]\fP"

.PP
Class constructor -- creates the instance but doesn't do anything\&. 
.PP
Definition at line \fB125\fP of file \fBAEFileWriter\&.hpp\fP\&.
.SS "AEFileWriter::AEFileWriter (\fBAEFileWriter\fP && fw)\fR [noexcept]\fP"

.PP
Move constructor\&. 
.PP
\fBParameters\fP
.RS 4
\fIfw\fP Object to be moved
.RE
.PP

.PP
Definition at line \fB20\fP of file \fBAEFileWriter\&.cpp\fP\&.
.SS "AEFileWriter::AEFileWriter (const \fBAEFileWriter\fP &)\fR [delete]\fP"

.PP
Deleted copy constructor\&. There is no need to copy AEFW, since access to file is in instance's FILE pointer\&. If in original instance, the file gets closed, the pointer is invalidated\&. Which can lead to\&.\&.\&.bad consequences using it again in the copied instance\&. 
.SS "AEFileWriter::~AEFileWriter (void)\fR [inline]\fP, \fR [noexcept]\fP"

.PP
Default Destructor\&. Just flushes and closes the file\&. 
.PP
Definition at line \fB153\fP of file \fBAEFileWriter\&.hpp\fP\&.
.PP
References \fBcloseFile()\fP, and \fBflushFile()\fP\&.
.SH "Member Function Documentation"
.PP 
.SS "\fBAEFileWriter\fP & AEFileWriter::operator= (const \fBAEFileWriter\fP &)\fR [delete]\fP"

.PP
Deleted copy assignment operator\&. There is no need to copy AEFW, since access to file is in instance's FILE pointer\&. If in original instance, the file gets closed, the pointer is invalidated\&. Which can lead to\&.\&.\&.bad consequences using it again in the copied instance\&. 
.SS "\fBcint\fP AEFileWriter::openFile (const std::string_view str, const \fBcint\fP flags = \fR0\fP, const \fBullint\fP af_interval = \fR\fBAEFW_DEFAULT_AUTOFLUSH_INTERVAL\fP\fP)"

.PP
Open(and create) file with given name and flag\&. On success sets \fBAEFileWriter::m_cFlags\fP to passed flag value\&. 
.PP
\fBNote\fP
.RS 4
If the provided file flag is invalid, it returns AEFW_ERR_OPEN_FILE_WRONG_FLAG 
.PP
If provided autoflush interval is a default value (AEFW_DEFAULT_AUTOFLUSH_INTERVAL), then the interval is not changed from the previous value 
.RE
.PP
\fBWarning\fP
.RS 4
Using the AEFW_FLAG_APPEND_NO_CURSOR_MOVE disables \fBAEFileWriter::getCursorPos()\fP, \fBAEFileWriter::setCursorPos()\fP, and \fBAEFileWriter::getFileSize()\fP, since they manipulate the cursor\&.
.RE
.PP
\fBParameters\fP
.RS 4
\fIstr\fP Name of the file, with extension
.br
\fIflags\fP Flags for file opening, AEFW_FLAG_* macros\&. More info in the docs
.br
\fIaf_interval\fP Interval in file writes between automatic file flushing\&.
.RE
.PP
\fBReturns\fP
.RS 4
AEFW_ERR_NOERROR if file was opened successfully; otherwise AEFW_ERR_OPEN_* flags (like AEFW_ERR_OPEN_FILE_NAME_EMPTY)
.RE
.PP

.PP
Definition at line \fB58\fP of file \fBAEFileWriter\&.cpp\fP\&.
.PP
References \fBAEFW_DEFAULT_AUTOFLUSH_INTERVAL\fP, \fBAEFW_ERR_NOERROR\fP, \fBAEFW_ERR_OPEN_FILE_ALREADY_OPENED\fP, \fBAEFW_ERR_OPEN_FILE_ELSE\fP, \fBAEFW_ERR_OPEN_FILE_NAME_EMPTY\fP, \fBAEFW_ERR_OPEN_FILE_WRONG_FLAG\fP, \fBAEFW_FLAG_APPEND\fP, \fBAEFW_FLAG_APPEND_NO_CURSOR_MOVE\fP, \fBAEFW_FLAG_NOFLAGS\fP, \fBAEFW_FLAG_TRUNCATE\fP, \fBace::utils::fopenCC()\fP, \fBisClosed()\fP, \fBisOpen()\fP, \fBm_cFlags\fP, \fBm_fpFilestr\fP, \fBm_sFilename\fP, \fBm_szLastWrittenAmount\fP, and \fBm_ullFlushInterval\fP\&.
.SS "\fBcint\fP AEFileWriter::flushFile (void) const\fR [inline]\fP, \fR [noexcept]\fP"

.PP
Flushes the opened file\&. That's it\&.
.PP
\fBReturns\fP
.RS 4
AEFW_ERR_WRITE_SUCCESS on success; otherwise AEFW_ERR_FILE_NOT_OPEN if file isn't open, AEFW_ERR_FLUSH_ERROR on flush error
.RE
.PP

.PP
Definition at line \fB177\fP of file \fBAEFileWriter\&.hpp\fP\&.
.PP
References \fB_AEFW_EXIT_ON_CLOSED_FILE\fP, \fBAEFW_ERR_FLUSH_ERROR\fP, \fBAEFW_ERR_WRITE_SUCCESS\fP, and \fBm_fpFilestr\fP\&.
.SS "\fBcint\fP AEFileWriter::closeFile (void)\fR [inline]\fP, \fR [noexcept]\fP"

.PP
Closes the currently opened file\&. 
.PP
\fBReturns\fP
.RS 4
AEFW_ERR_NOERROR if file was closed successfully; otherwise AEFW_ERR_FILE_NOT_OPEN if file isn't open
.RE
.PP

.PP
Definition at line \fB191\fP of file \fBAEFileWriter\&.hpp\fP\&.
.PP
References \fB_AEFW_EXIT_ON_CLOSED_FILE\fP, \fBAEFW_ERR_NOERROR\fP, \fBm_fpFilestr\fP, \fBm_sFilename\fP, and \fBm_szLastWrittenAmount\fP\&.
.SS "template<typename T > \fBcint\fP AEFileWriter::write (const T & var, const size_t datasz = \fR0\fP, const bool useAutoFlush = \fR\fBAEFW_DEFAULT_AUTOFLUSH\fP\fP)\fR [inline]\fP, \fR [noexcept]\fP"

.PP
Generic function for writing data, invoking proper write functions for built-in types\&. 
.PP
\fBNote\fP
.RS 4
String types don't include null-termination characters\&. Use separate functions for to control that 
.PP
Supported types: char, bool, integers, floats, strings; everything else is treated as binary stream\&.
.RE
.PP
\fBTemplate Parameters\fP
.RS 4
\fIT\fP Type of the variable to be written
.RE
.PP
\fBParameters\fP
.RS 4
\fIvar\fP Variable/data piece to be written
.br
\fIdatasz\fP Size of the data, in bytes\&. Only used if the T is a pointer to a binary stream, then it must be non-zero
.br
\fIuseAutoFlush\fP Flag to use automatic file flushing each n writes, specified by m_ullFlushInterval
.RE
.PP
\fBReturns\fP
.RS 4
AEFW_ERR_WRITE_SUCCESS on success; otherwise AEFW_ERR_FILE_NOT_OPEN if file isn't open, AEFW_ERR_WRITE_* flags (like AEFW_ERR_WRITE_ZERO_SIZE) on write error, AEFW_ERR_FLUSH_ERROR on flush error
.RE
.PP
\fBTodo\fP
.RS 4
Add a way to add custom types to this generic write function\&. 
.RE
.PP

.PP
Definition at line \fB623\fP of file \fBAEFileWriter\&.hpp\fP\&.
.PP
References \fB_AEFW_EXIT_ON_WRITE_CLOSED_FILE\fP, and \fBIS_SAME_NOCV\fP\&.
.SS "\fBcint\fP AEFileWriter::writeString (const std::string & str, const bool includeNull = \fRfalse\fP, const bool useAutoFlush = \fR\fBAEFW_DEFAULT_AUTOFLUSH\fP\fP)\fR [inline]\fP, \fR [noexcept]\fP"

.PP
Write std::string, as is, to file, and flush if necessary\&. 
.PP
\fBParameters\fP
.RS 4
\fIstr\fP String to write
.br
\fIincludeNull\fP Flag to include the trailing null-terminating character in the string
.br
\fIuseAutoFlush\fP Flag to use automatic file flushing each n writes, specified by m_ullFlushInterval
.RE
.PP
\fBReturns\fP
.RS 4
AEFW_ERR_WRITE_SUCCESS on success; otherwise AEFW_ERR_FILE_NOT_OPEN if file isn't open, AEFW_ERR_WRITE_* flags (like AEFW_ERR_WRITE_ZERO_SIZE) on write error, AEFW_ERR_FLUSH_ERROR on flush error
.RE
.PP

.PP
Definition at line \fB226\fP of file \fBAEFileWriter\&.hpp\fP\&.
.PP
References \fB_AEFW_EXIT_ON_WRITE_CLOSED_FILE\fP, \fBAEFW_ERR_WRITE_ZERO_SIZE\fP, and \fBwriteData_ptr()\fP\&.
.SS "\fBcint\fP AEFileWriter::writeString (const std::vector< char > & str, const bool useAutoFlush = \fR\fBAEFW_DEFAULT_AUTOFLUSH\fP\fP)\fR [inline]\fP, \fR [noexcept]\fP"

.PP
Write std::vector<char> (string), as is, to file, and flush if necessary\&. 
.PP
\fBNote\fP
.RS 4
It just writes the whole vector to file (since I cannot determine it without good ol' slow strlen)\&. If you want to control the null-byte, pass the std::vector<char>\&.data() as the
.RE
.PP
\fBParameters\fP
.RS 4
\fIstr\fP String(in form of vector<char>) to write
.br
\fIuseAutoFlush\fP Flag to use automatic file flushing each n writes, specified by m_ullFlushInterval
.RE
.PP
\fBReturns\fP
.RS 4
AEFW_ERR_WRITE_SUCCESS on success; otherwise AEFW_ERR_FILE_NOT_OPEN if file isn't open, AEFW_ERR_WRITE_* flags (like AEFW_ERR_WRITE_ZERO_SIZE) on write error, AEFW_ERR_FLUSH_ERROR on flush error
.RE
.PP

.PP
Definition at line \fB244\fP of file \fBAEFileWriter\&.hpp\fP\&.
.PP
References \fB_AEFW_EXIT_ON_WRITE_CLOSED_FILE\fP, \fBAEFW_ERR_WRITE_ZERO_SIZE\fP, and \fBwriteData_ptr()\fP\&.
.SS "\fBcint\fP AEFileWriter::writeString (const char *const cdata, const bool includeNull = \fRfalse\fP, const bool useAutoFlush = \fR\fBAEFW_DEFAULT_AUTOFLUSH\fP\fP)\fR [inline]\fP, \fR [noexcept]\fP"

.PP
Write c-string, as is, to file, and flush if necessary\&. 
.PP
\fBParameters\fP
.RS 4
\fIcdata\fP С-String to write
.br
\fIincludeNull\fP Flag to include the null-terminating character at the end of the string
.br
\fIuseAutoFlush\fP Flag to use automatic file flushing each n writes, specified by m_ullFlushInterval
.RE
.PP
\fBReturns\fP
.RS 4
AEFW_ERR_WRITE_SUCCESS on success; otherwise AEFW_ERR_FILE_NOT_OPEN if file isn't open, AEFW_ERR_WRITE_* flags (like AEFW_ERR_WRITE_ZERO_SIZE) on write error, AEFW_ERR_FLUSH_ERROR on flush error
.RE
.PP

.PP
Definition at line \fB262\fP of file \fBAEFileWriter\&.hpp\fP\&.
.PP
References \fB_AEFW_EXIT_ON_WRITE_CLOSED_FILE\fP, \fBAEFW_ERR_WRITE_ZERO_SIZE\fP, and \fBwriteData_ptr()\fP\&.
.SS "\fBcint\fP AEFileWriter::writeInt (const \fBllint\fP num, const bool useAutoFlush = \fR\fBAEFW_DEFAULT_AUTOFLUSH\fP\fP)\fR [inline]\fP, \fR [noexcept]\fP"

.PP
Writes the (signed) integer as text to open file\&. 
.PP
\fBParameters\fP
.RS 4
\fInum\fP The signed integer to be written
.br
\fIuseAutoFlush\fP Flag to use automatic file flushing each n writes, specified by m_ullFlushInterval
.RE
.PP
\fBReturns\fP
.RS 4
AEFW_ERR_WRITE_SUCCESS on success; otherwise AEFW_ERR_FILE_NOT_OPEN if file isn't open, AEFW_ERR_WRITE_* flags (like AEFW_ERR_WRITE_ZERO_SIZE) on write error, AEFW_ERR_FLUSH_ERROR on flush error
.RE
.PP

.PP
Definition at line \fB286\fP of file \fBAEFileWriter\&.hpp\fP\&.
.PP
References \fB_AEFW_EXIT_ON_WRITE_CLOSED_FILE\fP, and \fBwriteString()\fP\&.
.SS "\fBcint\fP AEFileWriter::writeUInt (const \fBullint\fP num, const bool useAutoFlush = \fR\fBAEFW_DEFAULT_AUTOFLUSH\fP\fP)\fR [inline]\fP, \fR [noexcept]\fP"

.PP
Writes the (unsigned) integer as text to open file\&. 
.PP
\fBParameters\fP
.RS 4
\fInum\fP The unsigned integer to be written
.br
\fIuseAutoFlush\fP Flag to use automatic file flushing each n writes, specified by m_ullFlushInterval
.RE
.PP
\fBReturns\fP
.RS 4
AEFW_ERR_WRITE_SUCCESS on success; otherwise AEFW_ERR_FILE_NOT_OPEN if file isn't open, AEFW_ERR_WRITE_* flags (like AEFW_ERR_WRITE_ZERO_SIZE) on write error, AEFW_ERR_FLUSH_ERROR on flush error
.RE
.PP

.PP
Definition at line \fB300\fP of file \fBAEFileWriter\&.hpp\fP\&.
.PP
References \fB_AEFW_EXIT_ON_WRITE_CLOSED_FILE\fP, and \fBwriteString()\fP\&.
.SS "template<typename T > 
.br
requires (std::is_floating_point<T>::value == true)\fBcint\fP AEFileWriter::writeFloat (const T num, const bool useAutoFlush = \fR\fBAEFW_DEFAULT_AUTOFLUSH\fP\fP)\fR [inline]\fP, \fR [noexcept]\fP"

.PP
Writes the float value as text to the opened file\&. 
.PP
\fBTemplate Parameters\fP
.RS 4
\fIT\fP The floating point type of the variable
.RE
.PP
\fBParameters\fP
.RS 4
\fInum\fP The float number itself
.br
\fIuseAutoFlush\fP Flag to use automatic file flushing each n writes, specified by m_ullFlushInterval
.RE
.PP
\fBReturns\fP
.RS 4
AEFW_ERR_WRITE_SUCCESS on success; otherwise AEFW_ERR_FILE_NOT_OPEN if file isn't open, AEFW_ERR_WRITE_* flags (like AEFW_ERR_WRITE_ZERO_SIZE) on write error, AEFW_ERR_FLUSH_ERROR on flush error
.RE
.PP
\fBTodo\fP
.RS 4
Add custom way to format the float number 
.RE
.PP

.PP
Definition at line \fB598\fP of file \fBAEFileWriter\&.hpp\fP\&.
.PP
References \fB_AEFW_EXIT_ON_WRITE_CLOSED_FILE\fP, and \fBIS_SAME_NOCV\fP\&.
.SS "\fBcint\fP AEFileWriter::writeBool (const bool num, const bool useAutoFlush = \fR\fBAEFW_DEFAULT_AUTOFLUSH\fP\fP)\fR [inline]\fP, \fR [noexcept]\fP"

.PP
Writes the boolean to file as text (true/false)\&. 
.PP
\fBParameters\fP
.RS 4
\fInum\fP The bool to be written
.br
\fIuseAutoFlush\fP Flag to use automatic file flushing each n writes, specified by m_ullFlushInterval
.RE
.PP
\fBReturns\fP
.RS 4
AEFW_ERR_WRITE_SUCCESS on success; otherwise AEFW_ERR_FILE_NOT_OPEN if file isn't open, AEFW_ERR_WRITE_* flags (like AEFW_ERR_WRITE_ZERO_SIZE) on write error, AEFW_ERR_FLUSH_ERROR on flush error
.RE
.PP

.PP
Definition at line \fB329\fP of file \fBAEFileWriter\&.hpp\fP\&.
.PP
References \fBace::utils::boolToString()\fP, and \fBwriteString()\fP\&.
.SS "\fBcint\fP AEFileWriter::writeChar (const char c, const bool useAutoFlush = \fR\fBAEFW_DEFAULT_AUTOFLUSH\fP\fP)\fR [inline]\fP, \fR [noexcept]\fP"

.PP
Writes the single (ascii) character to the file\&. 
.PP
\fBNote\fP
.RS 4
Essentially, it's the same as \fBAEFileWriter::writeByte()\fP
.RE
.PP
\fBParameters\fP
.RS 4
\fIc\fP The char to be written
.br
\fIuseAutoFlush\fP Flag to use automatic file flushing each n writes, specified by m_ullFlushInterval
.RE
.PP
\fBReturns\fP
.RS 4
AEFW_ERR_WRITE_SUCCESS on success; otherwise AEFW_ERR_FILE_NOT_OPEN if file isn't open, AEFW_ERR_WRITE_* flags (like AEFW_ERR_WRITE_ZERO_SIZE) on write error, AEFW_ERR_FLUSH_ERROR on flush error
.RE
.PP

.PP
Definition at line \fB340\fP of file \fBAEFileWriter\&.hpp\fP\&.
.PP
References \fBwriteByte()\fP\&.
.SS "\fBcint\fP AEFileWriter::writeByte (const unsigned char cdata, const bool useAutoFlush = \fR\fBAEFW_DEFAULT_AUTOFLUSH\fP\fP)\fR [inline]\fP, \fR [noexcept]\fP"

.PP
Writes one byte of data to the file\&. 
.PP
\fBParameters\fP
.RS 4

AEFileWriter/src/AEFileWriter.cpp

/*! @file AEFileWriter/src/AEFileWriter.cpp
 *  This file contains the definitions of the engine's file writer's functions.
 *  @see AEFileWriter
 *  @see AEFileWriter.hpp
 *
 *  Should not cause everything to break :)
 */

#include "../include/AEFileWriter.hpp"

AEFileWriter::AEFileWriter(const std::string_view filename, const cint flags, const ullint af_interval) :
	m_ullFlushInterval(af_interval), m_ullTotalWrites(0), m_szLastWrittenAmount(0),
	 m_fpFilestr(nullptr), m_cFlags(flags) {

	this->openFile(filename, flags);
}


AEFileWriter::AEFileWriter(AEFileWriter&& fw) noexcept :
	m_ullFlushInterval(fw.m_ullFlushInterval), m_sFilename(std::move(fw.m_sFilename)),
	m_ullTotalWrites(fw.m_ullTotalWrites.load()), m_szLastWrittenAmount(fw.m_szLastWrittenAmount),
	m_fpFilestr(fw.m_fpFilestr), m_cFlags(fw.m_cFlags) {

	fw.m_ullFlushInterval = 0;
	fw.m_ullTotalWrites = 0;
	fw.m_szLastWrittenAmount = 0;
	fw.m_fpFilestr = nullptr;
	fw.m_sFilename.clear();
	fw.m_cFlags = 0;
}

// write data as binary function
// uses const char
cint AEFileWriter::writeData_ptr(const void* const cdata, const std::size_t dcount, const std::size_t dsize, const bool useAutoFlush) noexcept {

	_AEFW_EXIT_ON_WRITE_CLOSED_FILE;

	if (!cdata || !dcount || !dsize) {
		return AEFW_ERR_WRITE_ZERO_SIZE;
	}

	this->m_szLastWrittenAmount = fwrite(cdata, dsize, dcount, this->m_fpFilestr);
	this->m_ullTotalWrites.fetch_add(1, std::memory_order::relaxed);
	if (useAutoFlush) {
		this->autoFlush();
	}

	if (m_szLastWrittenAmount != dcount) {
		return AEFW_ERR_WRITE_ERROR;
	}

	return AEFW_ERR_WRITE_SUCCESS;
}


// open file with flags
cint AEFileWriter::openFile(const std::string_view str, const cint flags, const ullint af_interval) {

	if (this->isOpen()) { // open already -> pls close
		return AEFW_ERR_OPEN_FILE_ALREADY_OPENED;
	}

	if (str.empty()) {
		this->m_sFilename.clear();
		return AEFW_ERR_OPEN_FILE_NAME_EMPTY;
	}

	this->m_szLastWrittenAmount = 0;
	this->m_sFilename = str;
	std::replace(m_sFilename.begin(), m_sFilename.end(), '\\', '/');
	const std::size_t found = this->m_sFilename.rfind('/');
	if (found != std::string::npos) {
		std::filesystem::create_directories(this->m_sFilename.substr(0, found));
	}

	switch (flags) {

	case AEFW_FLAG_APPEND://cursor at end, allow changing cursor position
		this->m_fpFilestr = ace::utils::fopenCC(str.data(), "r+b");
		if (this->isClosed()) {//couldn't open, force create
			this->m_fpFilestr = ace::utils::fopenCC(str.data(), "wb");
			if (this->isClosed()) {
				break; // some other error
			}
		}
		fseek(this->m_fpFilestr, 0L, SEEK_END);
		break;

	case AEFW_FLAG_APPEND_NO_CURSOR_MOVE://only appending, not allowing to change cursor position
		this->m_fpFilestr = ace::utils::fopenCC(str.data(), "ab");
		break;


	case AEFW_FLAG_NOFLAGS://no flags, defaulting to truncation
	case AEFW_FLAG_TRUNCATE://truncate file
		this->m_fpFilestr = ace::utils::fopenCC(str.data(), "wb");
		break;

	default:
		return AEFW_ERR_OPEN_FILE_WRONG_FLAG;
	}

	// last check, to see if everything is okay
	if (this->isClosed()) {
		return AEFW_ERR_OPEN_FILE_ELSE;
	}

	//then everything is good, 
	this->m_cFlags = flags;
	if (this->m_ullFlushInterval != AEFW_DEFAULT_AUTOFLUSH_INTERVAL) {
		this->m_ullFlushInterval = af_interval;
	}
	return AEFW_ERR_NOERROR;
}
